warehouse/query-core/src/main/java/datawave/query/function/LimitFields.java

```diff
@@ -87,89 +88,127 @@ public Multimap<String,Attribute<? extends Comparable<?>>> load(String key) {
+                    increment(keyNoGrouping, countForFieldMap);
                 }
             }
         }
-        for (String keyNoGrouping : countForFieldMap.keySet()) {
-            int limit = this.limitFieldsMap.get(keyNoGrouping);
-            Multimap<String,Attribute<? extends Comparable<?>>> hitMap = hits.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : hitMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : hitMap.get(keyWithGrouping)) {
-                    // if(limit <= 0) break; // comment this line if you want to get ALL hits even if the limit is exceeded
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
-                }
+        
+        // second pass is to set any misses back to be kept if the limit allows
+        for (Map.Entry<String,Attribute<? extends Comparable<?>>> de : document.entrySet()) {
+            String keyWithGrouping = de.getKey();
+            String keyNoGrouping = keyWithGrouping;
+            // if we have grouping context on, remove the grouping context
+            if (keyNoGrouping.indexOf('.') != -1) {
+                keyNoGrouping = keyNoGrouping.substring(0, keyNoGrouping.indexOf('.'));
             }
-            Multimap<String,Attribute<? extends Comparable<?>>> missMap = misses.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : missMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : missMap.get(keyWithGrouping)) {
-                    if (limit <= 0)
-                        break;
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
+            
+            if (this.limitFieldsMap.containsKey(keyNoGrouping)) { // look for the key without the grouping context
+                int limit = this.limitFieldsMap.get(keyNoGrouping); // used below if you un-comment to get all hits
+                
+                // short circuit if we are not actually limiting this field.
+                // this is keeping with the original logic where a negative limit means to keep only hits
+                if (limit < 0) {
+                    continue;
+                }
+                
+                int missesToSet = Math.min(limit - get(keyNoGrouping, countKeepersForFieldMap), get(keyNoGrouping, countMissesRemainingForFieldMap));
+                
+                // if we have misses yet to keep
+                if (missesToSet > 0) {
+                    
+                    Attribute<?> attr = de.getValue();
+                    if (attr instanceof Attributes) {
+                        Attributes attrs = (Attributes) attr;
+                        Set<Attribute<? extends Comparable<?>>> attrSet = attrs.getAttributes();
+                        
+                        for (Attribute<? extends Comparable<?>> value : attrSet) {
+                            // if this was a miss previously set to not keep
```

Comment:
```suggestion
                            // if this was a miss previously set to keep now
```

Reply:
The sentence was actually correct.  Rewording to me more clear however.